### ISSUE TITLE
Feature request: force download of all files in an  iCloud folder, including hidden files

### ISSUE THREAD

**Issue by chchen2021** [holzschu/a-shell#300]
Is it possible to force iCloud to download an entire folder, including hidden files? For example, can we include a command in `~/Documents/.profile` so all files in a-Shell iCloud folder are automatically downloaded upon opening a-Shell?

When I use macOS and a-Shell to manage the same iCloud folder (e.g. with pdflatex or task), I often encounter the annoyance of iCloud files not yet downloaded locally (so the files have an .icloud suffix). So far the best solution is to select all files and hit copy, which will force the downloading of all visible files in a folder, but this solution has two problems:

1. I need to leave the CLI and use the Files app to select all files and hit copy. I have tried commands like `open .` but it does not open the current directory in Files, unlike in macOS.
2. There is no way to select hidden folders or hidden files in the Files app. I ended up renaming `.task` to `task` just so that I can select the files in the folder to force download. Such renaming works with a modification to `.taskrc`, but would not work in some scenarios (e.g. the `.obsidian` file within a Obsidian vault).

So I am basically asking for an iCloud download behavior similar to what I see in macOS, in which files are automatically downloaded to the local disk.

**Comment by holzschu** [comment]
I see what you mean. I agree that it's really annoying. I'm a bit wary of doing things automatically (after `ls -a`, for example), as users might not always want to start a download (due to poor network connection, for example). But selecting a folder in `pickFolder` should definitely result in downloading all the files from that folder.

I'm adding code for that in a-Shell, with a time limit (if the network connection is bad, a-Shell will try downloading files for 5 seconds before giving up).

**Comment by chchen2021** [comment]
Thanks for being so receptive to the feature requests. I would like to offer some additional considerations:

- If a user does not want to automatically download a folder, they may not want it with pickFolder either. Even worse, if a bookmark is persistent, they may forget about the automatic download in future, and put a really big file in the bookmarked folder that ends up using wireless data.
- It's not ideal to use pickFolder to force download a-Shell's own iCloud folder, `~iCloud` (`~cloud`). It hurts the workflow, and also creates a third bookmark, `~Documents`, which can be very confusing. 

My recommendation is to have an additional command like "downloadFolder". One can either use it in the command line, or include it in `~/Documents/.profile` so an included folder is always downloaded. 

I am not sure about the 5-second rule. The problem with non-yet-downloaded iCloud files is that the names of the files are occupied, and a new file will often be created (so "file 1" instead of "file"). I am not experienced in iCloud behavior to make any useful suggestion though.

**Comment by holzschu** [comment]
You're right. I will add the `downloadFolder` and `downloadFile` command. Still, if I pick a folder inside iCloud using pickFolder, I definitely expect the files inside to be available, so I'll also keep the recent addition (not just a-Shell's own iCloud folder, it can be somewhere else on iCloud).

I've had some bad experiments with iCloud over the summer, including the `iCloud/Documents` folder being unavailable for weeks, so I would advise you to keep local copies of anything important.

**Comment by chchen2021** [comment]
Thanks for the cautionary note. My `~/Library/Mobile Documents` folder is backed up by Time Machine, and the backup has indeed saved me from losing an entire iCloud folder. As long as there is sufficient hard drive space, I think all iCloud files are automatically downloaded by macOS, so Time Machine is a good choice for backup.

**Comment by holzschu** [comment]
No, I can confirm it's an a-Shell issue, related to more security around commands launched by other commands (to avoid having one command changing the environment and directory of another). It will be fixed in the next build.

**Comment by holzschu** [comment]
Hi, this should be fixed with build 203 currently on TestFlight ( https://testflight.apple.com/join/WUdKe3f4 ). If you can confirm that it works for you, I will release it for everyone.

**Comment by chchen2021** [comment]
I can confirm the `find -exec` bug is gone. Thanks for fixing it. 

Also, the original latexmk works, but somehow it is no longer included in the last few builds. I just copied on in my bin folder and it works fine.

**Comment by chchen2021** [comment]
latexmk is indeed included. 

iIt was an easy reinstallation on the iPad through Settings, but on the iPhone, I ended up removing the entire a-Shell to avoid repeated crashes. Everything is working fine now.
